aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -232,7 +278,11 @@ namespace {
 
 SparseTensor& resize_as_sparse_(SparseTensor& self, const SparseTensor& src) {
   if (!_is_same_size_as_sparse(self, src)) {
-    _raw_resize_sparse(self, src._sparseDims(), src._denseDims(), src.sizes());
+#ifndef USE_TH_SIZE_ZERO_DIM
+    _raw_resize_sparse_legacy(self, src._sparseDims(), src._denseDims(), src.sizes());
```

Comment:
For the new semantics the function should probably be named `resize_as_sparse_and_clear_`, but then it would be weird for it to still be in the `resize_as_` Python code path because it now does both resize and clear. To resolve this, can we assume that the user won't care about the internal data when calling `resize_as_` on a sparse tensor (and update our doc to explain the new resize and clear behavior)?

Reply:
no, I don't think that is a valid assumption because it doesn't match the semantics of dense tensors.  Where is this function actually used?

Reply:
For the semantics of `resize_as_`, I wondered if we can break it down into two cases:

1. Different size, but same # of elements (e.g. `torch.Size([1, 2, 3])` and `torch.Size([2, 3])`). sparseDims/denseDims can be same or different.
In this case, we will convert the data in `indices` and `values` to the new size and sparseDims/denseDims. All data are preserved in this case.

2. Different size, and different # of elements
In this case, since it's difficult to know where the original data will be mapped to in the new resized tensor (the mapping is probably very implementation-specific; we could run `to_dense()` -> `resize_()` -> `nonzero()` to get the new indices, but it would be very memory inefficient), I suggest that we give a UserWarning in this case saying that the # of elements are different and hence data can't be preserved, and we just call `resize_and_clear` to get an empty sparse tensor with the new size in this case.